crates/sui-indexer-alt-framework/src/pipeline/concurrent/commit_watermark.rs

```diff
@@ -255,3 +255,303 @@ pub(super) fn commit_watermark<H: Handler + 'static>(
+                .unwrap()
+                .unwrap();
+        assert_eq!(watermark.checkpoint_hi_inclusive, 6);
+
+        // Cancel and wait for shutdown
+        cancel.cancel();
+        handle.await.unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_commit_watermark_skip() {
+        let (_temp_db, db) = test_db().await;
+        let registry = Registry::new_custom(Some("indexer_alt".to_string()), None).unwrap();
+        let metrics = Arc::new(IndexerMetrics::new(&registry));
+        let cancel = CancellationToken::new();
+        let (_tx, rx) = mpsc::channel(100);
+
+        let handle = commit_watermark::<TestHandler>(
+            None,
+            CommitterConfig::default(),
+            true,
+            rx,
+            db,
+            metrics,
+            cancel,
+        );
+
+        // Task should exit immediately when skip_watermark is true
+        handle.await.unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_commit_watermark_out_of_order() {
+        let (_temp_db, db) = test_db().await;
+        let registry = Registry::new_custom(Some("indexer_alt".to_string()), None).unwrap();
+        let metrics = Arc::new(IndexerMetrics::new(&registry));
+        let cancel = CancellationToken::new();
+        let (tx, rx) = mpsc::channel(100);
+
+        let handle = commit_watermark::<TestHandler>(
+            None,
+            CommitterConfig::default(),
+            false,
+            rx,
+            db.clone(),
+            metrics.clone(),
+            cancel.clone(),
+        );
+
+        // Send a correct watermark part first
+        let part1 = WatermarkPart {
+            watermark: new_commit_watermark(0).await,
+            batch_rows: 10,
+            total_rows: 10,
+        };
+        tx.send(vec![part1]).await.unwrap();
+        // Let the task process the watermark
+        tokio::time::sleep(Duration::from_millis(50)).await;
+        let watermark =
+            CommitterWatermark::read_highest_watermark(&mut db.connect().await.unwrap(), "test")
+                .await
+                .unwrap()
+                .unwrap();
+        assert_eq!(watermark.checkpoint_hi_inclusive, 0);
+
+        // Send an out of order watermark part
+        let part2 = WatermarkPart {
+            watermark: new_commit_watermark(2).await,
+            batch_rows: 10,
+            total_rows: 10,
+        };
+        tx.send(vec![part2]).await.unwrap();
+
+        // Let the task process the watermark
+        tokio::time::sleep(Duration::from_millis(50)).await;
+        // Verify watermark not updated since checkpoint 1 has not been received yet.
+        let watermark =
+            CommitterWatermark::read_highest_watermark(&mut db.connect().await.unwrap(), "test")
+                .await
+                .unwrap()
+                .unwrap();
+        assert_eq!(watermark.checkpoint_hi_inclusive, 0);
+
+        // Send the correct watermark part for checkpoint 1
+        let part1 = WatermarkPart {
```

Comment:
nit: numbering is a little confusing here (watermark 0 is part1, then 2 is part2, then 1 is part1 again)